modules/docs/src/05-design/03-services.md

```diff
@@ -0,0 +1,293 @@
+
+## The Endpoint abstraction
+
+The `smithy4s.Endpoint` abstraction ties a specific operation to the various schemas that are tied to it.
+
+```scala
+trait Endpoint[Initial[_, _, _, _, _], I, E, O, SI, SO] {
+  def shapeId : ShapeId
+  def hints: Hints
+  def input: Schema[I]
+  def output: Schema[O]
+  def streamedInput: StreamingSchema[SI]
+  def streamedOutput: StreamingSchema[SO]
+  def wrap(input: I): Initial[I, E, O, SI, SO]
+
+  def errorable: Option[Errorable[E]]
+}
+```
+
+Endpoints are not typeclasses. Instead, the `Endpoint` trait is extended by the companion object of each member of the `GADT` forming the initial-encoding of the service interface. So, going back to our `KVStore`, the corresponding sealed-trait would look like this :
+
+```scala
+sealed trait KVStoreOp[Input, Error, Output, StreamedInput, StreamedOutput]
+```
+
+and the `put` operation would look like :
+
+```scala
+case class Put(input: PutRequest) extends KVStoreOp[PutRequest, PutError, PutResult, Nothing, Nothing]
+object Put extends Endpoint[KVStoreOp, PutRequest, PutError, PutResult, Nothing, Nothing] with Errorable[PutError]{
+  val input = PutRequest.input
+  val output = PutRequest.input
+  val streamedInput = SteamingSchema.nothing
+  val streamedOutput = StreamingSchema.nothing
+  val errorable: Option[Errorable[PutResult]] = this
+  // ...
+}
+```
+
+### A note on errors
+
+As stated previously, smithy4s generates a coproduct type for each operation, where the members of the coproduct point to the various errors listed in the smithy operation shape. Additionally, each structure annotated with `@error` in smithy is rendered as a case-class that extends `Throwable`, because `Throwables` are the de-facto standard of doing error handling on the JVM. Even libraries that use `Either` to perform error handling often represent the left-hand-side of the Either as some throwable type, to facilitate the absorption of errors into the error-channels of monadic constructs (`IO.raiseError`, etc)
+
+As a result, it is important for smithy4s to expose functions that generically enable the filtering of throwables against the `Error` type parameter of an operation, so that interpreters can intercept errors and apply the correct encoding (dictated via `Schema`) before communicating them back to the caller over the user. Conversely, it is important to expose a function that allows to go from the generic `Error` type parameter to `Throwable`, so that errors received via low-level communication channels can be turnd into `Throwable` at the client call site, in order to populate the relevant error channel when exposing monofunctor semantics.
```

Comment:
```suggestion
As a result, it is important for smithy4s to expose functions that generically enable the filtering of throwables against the `Error` type parameter of an operation, so that interpreters can intercept errors and apply the correct encoding (dictated via `Schema`) before communicating them back to the caller over the user. Conversely, it is important to expose a function that allows to go from the generic `Error` type parameter to `Throwable`, so that errors received via low-level communication channels can be turned into `Throwable` at the client call site, in order to populate the relevant error channel when exposing monofunctor semantics.
```